src/lib/support/FluentTreeObject.h

```diff
@@ -0,0 +1,137 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <lib/support/Span.h>
+
+#include <optional>
+
+namespace chip {
+
+/// represents a wrapper around a type `T` that contains internal
+/// `Span<...>` values of other sub-types. It allows searching within the container sub-spans
+/// to create new containers.
+///
+/// Use case is that we very often search within a tree, like "find-endpoint" + "find-cluster" + "find-attribute"
+/// and we generally only care if "does the last element exist or not"
+///
+/// General usage is for fluent-searching for things like:
+///
+///    FluentTreeObject container(somePointer);
+///
+///    const AcceptedCommandData * value =
+///           container
+///              .Find<ByEndpoint>(path.mEndpointId, mEndpointIndexHint)
+///              .Find<ByServerCluster>(path.mClusterId, mServerClusterHint)
+///              .Find<ByAcceptedCommand>(path.mCommandId, mAcceptedCommandHint)
+///              .Value();
+///
+/// Where a `ByFoo` structure looks like:
+///
+///    struct ByFoo {
+///      using Key  = int;                  // the KEY inside a type
+///      using Type = SomeContainerStruct;  // the values for a sub-search
+///      static Span<Type> GetSpan(Data & data) { /* return ... */ }
+///      static bool HasKey(const Key & id, const Type & instance) { /* return "instance has key id" */ }
+///    }
+///
+/// Where we define:
+///    - how to get a "span of sub-elements" for an object (`GetSpan`)
+///    - how to determine if a given sub-element has the "correct key"
+template <typename T>
+class FluentTreeObject
+{
+public:
+    explicit FluentTreeObject(T * value) : mValue(value) {}
+
+    /// Returns NULLPTR if such an element does not exist or non-null valid value if the element exists
+    T * Value() const { return mValue; }
+
+    // Get the first element of `TYPE`
+    template <typename TYPE>
+    FluentTreeObject<typename TYPE::Type> First(unsigned & indexHint)
+    {
+        // if no value, searching more also yields no value
+        VerifyOrReturnValue(mValue != nullptr, FluentTreeObject<typename TYPE::Type>(nullptr));
```

Comment:
I wonder whether defining a protected non-explicit constructor that takes nullptr_t would make this easier to write.  As in, your could then have `VerifyOrReturnValue(mValue != nullptr, nullptr);`. Or is that too magic?

Reply:
Looks cleaner - updated to have default constructor and nullptr constructor allowed for null.